Input checking

christophercolumbusdog opened the pull request:
Validate input on the RegisterActivity when a user is trying to sign up

christophercolumbusdog committed:
Check password format

christophercolumbusdog committed:
Added basic checks and masking

christophercolumbusdog committed:
Added basic checks and masking

christophercolumbusdog committed:
Add email regex

christophercolumbusdog committed:
Add phone number regex

christophercolumbusdog committed:
added birthday regex

christophercolumbusdog committed:
Added full error checks on input

christophercolumbusdog committed:
Added full error checks on input

christophercolumbusdog committed:
Refactor to fix checkstyle

christophercolumbusdog committed:
Add birthday transform to YYYY-MM-DD format in database

pgalusza commented on app/src/main/java/com/example/mac/chartr/activities/RegisterActivity.java:
```diff
@@ -326,15 +497,33 @@ private void initEmail() {
+                                + "[(?:(?:25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)\\.){3}(?:25[0-5]"
+                                + "|2[0-4][0-9]|[01]?[0-9][0-9]?|[a-z0-9-]*[a-z0-9]:(?:[\\x01-\\"
+                                + "x08\\x0b\\x0c\\x0e-\\x1f\\x21-\\x5a\\x53-\\x7f]|\\\\[\\x01-\\"
+                                + "x09\\x0b\\x0c\\x0e-\\x7f])+)\\])");
```
The regular expression is letting through invalid emails with multiple @, such as a@b@c@d.com.

christophercolumbusdog replied:
Hm, yeah I'm not sure the best way to address it. Currently I based that off of: http://emailregex.com

If we can find a more precise or restrictive one, that could work. I'm not sure, what are your thoughts?

pgalusza replied:
Just figured it out.  It was finding patterns that did not necessarily start at the beginning of the string.  If we add a "^" at the start of the pattern, then the matched pattern has to start at the beginning of the string.  Other than that, do you want to allow for white space before the email, or not?

christophercolumbusdog replied:
Ahhhh, ok, good catch! Uh, I'd say to not allow whitespace before the email

pgalusza committed:
Update password checker

Password checker updated to accept all the valid special characters as specified by AWS:
^ $ * . [ ] { } ( ) ? - " ! @ # % & / \ , > < ' : ; | _ ~ `

pgalusza commented on app/src/main/java/com/example/mac/chartr/activities/RegisterActivity.java:
```diff
@@ -208,23 +291,41 @@ private void initSignup() {
+                } else {
+                    message.setText("Invalid phone number");
+                }
             }
```
I really don't like how the phone input looks, is it ok if I change the phone related input so that it shows up as either:
+1-222-333-4444 or
+1 (222) 333-4444
in the input?  If so, which do you prefer?

christophercolumbusdog replied:
I agree that the current input does look rather bad. I would say +1 (222) 333-4444 is the ideal format. However, we'd just have to modify the regex for that. But honestly, that's not a complex pattern to recognize. Also look into the Mask Watcher which can do some of the auto formatting.

On thing though: the whole 
"+1 (222) 333-4444" is an American format, so the one thing is that it wouldn't really support the international formats. Currently, I'm doing a regex and checking that will allow different international format (only enforcing a specific US format if it starts with a 1). Now, we don't necessarily need to support all country phone numbers, but it's something to consider.

pgalusza replied:
I can change it so that it enforces the phone format the same way as it does the date.  It doesn't look too complicated to do if we stick with one format.

pgalusza replied:
Since we're assuming the length is 11 digits anyway, I'll just use the +1 (222) 333-4444 format for now.

pgalusza commented on app/src/main/java/com/example/mac/chartr/activities/RegisterActivity.java:
```diff
@@ -151,48 +161,121 @@ private void initSignup() {
 
+            /*
+             Check name and add to AppHelper Attribute
+             */
```
Aren't the following checks redundant after you put in the new ones?

christophercolumbusdog replied:
Yeah, we actually should be able to remove the "if", just keep the logic because if it makes it that far, the strings are going to be non-empty

pgalusza committed:
Changed input formatting and checking.

Phone input now formatted as "+1-222-333-4444" and email input no longer accepts emails with more than one "@".  Also removed redundant code.

pgalusza reviewed (approved):
Input checking for register user looks good now.
